File: client/csharp/Angzarr.Client/CloudEvents.cs
using System.Reflection;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Type = System.Type;

namespace Angzarr.Client;

/// <summary>
/// CloudEvents projectors transform internal domain events into CloudEvents 1.0 format
/// for external consumption via HTTP webhooks or Kafka.
///
/// OO Pattern (CloudEventsProjector):
/// <code>
/// public class PlayerCloudEventsProjector : CloudEventsProjector
/// {
///     public override string Name => "prj-player-cloudevents";
///     public override string InputDomain => "player";
///
///     [CloudEventsHandler(typeof(PlayerRegistered))]
///     public CloudEvent? OnPlayerRegistered(PlayerRegistered evt) =>
///         new CloudEvent
///         {
///             Type = "com.poker.player.registered",
///             Data = Any.Pack(new PublicPlayerRegistered { DisplayName = evt.DisplayName })
///         };
/// }
/// </code>
///
/// Functional Pattern (CloudEventsRouter):
/// <code>
/// var router = new CloudEventsRouter("prj-player-cloudevents", "player")
///     .On&lt;PlayerRegistered&gt;(evt => new CloudEvent
///     {
///         Type = "com.poker.player.registered",
///         Data = Any.Pack(new PublicPlayerRegistered { DisplayName = evt.DisplayName })
///     });
/// </code>
/// </summary>
public abstract class CloudEventsProjector
{
    private static readonly Dictionary<
        Type,
        Dictionary<string, (MethodInfo Method, Type EventType)>
    > _dispatchTables = new();

    /// <summary>
    /// The name of this projector (e.g., "prj-player-cloudevents").
    /// </summary>
    public abstract string Name { get; }

    /// <summary>
    /// The domain this projector listens to.
    /// </summary>
    public abstract string InputDomain { get; }

    protected CloudEventsProjector()
    {
        EnsureDispatchTableBuilt();
    }

    private void EnsureDispatchTableBuilt()
    {
        var type = GetType();
        if (_dispatchTables.ContainsKey(type))
            return;

        lock (_dispatchTables)
        {
            if (_dispatchTables.ContainsKey(type))
                return;

            var dispatch = new Dictionary<string, (MethodInfo, Type)>();
            foreach (
                var method in type.GetMethods(
                    BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic
                )
            )
            {
                var attr = method.GetCustomAttribute<CloudEventsHandlerAttribute>();
                if (attr != null)
                {
                    var suffix = attr.EventType.Name;
                    dispatch[suffix] = (method, attr.EventType);
                }
            }
            _dispatchTables[type] = dispatch;
        }
    }

    /// <summary>
    /// Process an EventBook and return CloudEvents.
    /// </summary>
    public static CloudEventsResponse Handle<T>(Angzarr.EventBook source)
        where T : CloudEventsProjector, new()
    {
        var projector = new T();
        return projector.Project(source);
    }

    /// <summary>
    /// Process an EventBook and return CloudEvents.
    /// </summary>
    public CloudEventsResponse Project(Angzarr.EventBook source)
    {
        var response = new CloudEventsResponse();
        var dispatchTable = _dispatchTables[GetType()];

        foreach (var page in source.Pages)
        {
            if (page.Event == null)
                continue;

            foreach (var (suffix, (method, eventType)) in dispatchTable)
            {
                if (page.Event.TypeUrl.EndsWith(suffix))
                {
                    var unpackMethod = typeof(Any)
                        .GetMethod("Unpack")!
                        .MakeGenericMethod(eventType);
                    var evt = unpackMethod.Invoke(page.Event, null);
                    var result = method.Invoke(this, new[] { evt });
                    if (result is CloudEvent ce)
                    {
                        response.Events.Add(ce);
                    }
                    break;
                }
            }
        }

        return response;
    }

    /// <summary>
    /// Handle a projector request, returning a Projection containing CloudEventsResponse.
    /// </summary>
    public Angzarr.Projection HandleRequest(Angzarr.EventBook source)
    {
        var response = Project(source);
        return new Angzarr.Projection
        {
            Projector = Name,
            Projection_ = Any.Pack(response, "type.googleapis.com/"),
        };
    }
}

/// <summary>
/// Attribute to mark a method as a CloudEvents handler.
/// </summary>
[AttributeUsage(AttributeTargets.Method)]
public class CloudEventsHandlerAttribute : Attribute
{
    public Type EventType { get; }

    public CloudEventsHandlerAttribute(Type eventType)
    {
        EventType = eventType;
    }
}

/// <summary>
/// Functional router for CloudEvents projectors.
///
/// Example:
/// <code>
/// var router = new CloudEventsRouter("prj-player-cloudevents", "player")
///     .On&lt;PlayerRegistered&gt;(evt => new CloudEvent
///     {
///         Type = "com.poker.player.registered",
///         Data = Any.Pack(new PublicPlayerRegistered { DisplayName = evt.DisplayName })
///     });
///
/// var response = router.Project(eventBook);
/// </code>
/// </summary>
public class CloudEventsRouter
{
    private readonly string _name;
    private readonly string _inputDomain;
    private readonly List<CloudEventsEntry> _handlers = new();

    private sealed class CloudEventsEntry
    {
        public string Suffix { get; }
        public Type EventType { get; }
        public Delegate Handler { get; }

        public CloudEventsEntry(string suffix, Type eventType, Delegate handler)
        {
            Suffix = suffix;
            EventType = eventType;
            Handler = handler;
        }
    }

    /// <summary>
    /// Create a new CloudEvents router.
    /// </summary>
    /// <param name="name">The projector name.</param>
    /// <param name="inputDomain">The domain this projector listens to.</param>
    public CloudEventsRouter(string name, string inputDomain)
    {
        _name = name;
        _inputDomain = inputDomain;
    }

    /// <summary>
    /// The projector name.
    /// </summary>
    public string Name => _name;

    /// <summary>
    /// The domain this projector listens to.
    /// </summary>
    public string InputDomain => _inputDomain;

    /// <summary>
    /// Register a handler for an event type.
    ///
    /// The handler receives the typed event and returns a CloudEvent or null to skip.
    /// </summary>
    /// <typeparam name="TEvent">The event type to handle.</typeparam>
    /// <param name="handler">Function that transforms event to CloudEvent.</param>
    /// <returns>This router for fluent chaining.</returns>
    public CloudEventsRouter On<TEvent>(Func<TEvent, CloudEvent?> handler)
        where TEvent : IMessage
    {
        var suffix = typeof(TEvent).Name;
        _handlers.Add(new CloudEventsEntry(suffix, typeof(TEvent), handler));
        return this;
    }

    /// <summary>
    /// Register a handler for an event type suffix.
    ///
    /// The handler receives the Any payload and returns a CloudEvent or null to skip.
    /// </summary>
    /// <param name="suffix">The type_url suffix to match.</param>
    /// <param name="handler">Function that transforms Any to CloudEvent.</param>
    /// <returns>This router for fluent chaining.</returns>
    public CloudEventsRouter OnSuffix(string suffix, Func<Any, CloudEvent?> handler)
    {
        _handlers.Add(new CloudEventsEntry(suffix, typeof(Any), handler));
        return this;
    }

    /// <summary>
    /// Process an EventBook and return CloudEvents.
    /// </summary>
    /// <param name="source">The EventBook containing events to process.</param>
    /// <returns>CloudEventsResponse containing produced CloudEvents.</returns>
    public CloudEventsResponse Project(Angzarr.EventBook source)
    {
        var response = new CloudEventsResponse();

        foreach (var page in source.Pages)
        {
            if (page.Event == null)
                continue;

            var typeUrl = page.Event.TypeUrl;

            foreach (var entry in _handlers)
            {
                if (typeUrl.EndsWith(entry.Suffix))
                {
                    CloudEvent? result;

                    if (entry.EventType == typeof(Any))
                    {
                        // Suffix handler - pass Any directly
                        result = ((Func<Any, CloudEvent?>)entry.Handler)(page.Event);
                    }
                    else
                    {
                        // Typed handler - unpack and invoke
                        var unpackMethod = typeof(Any)
                            .GetMethod("Unpack")!
                            .MakeGenericMethod(entry.EventType);
                        var evt = unpackMethod.Invoke(page.Event, null);
                        result = (CloudEvent?)entry.Handler.DynamicInvoke(evt);
                    }

                    if (result != null)
                    {
                        response.Events.Add(result);
                    }
                    break;
                }
            }
        }

        return response;
    }

    /// <summary>
    /// Handle a projector request, returning a Projection containing CloudEventsResponse.
    /// </summary>
    /// <param name="source">The EventBook containing events to process.</param>
    /// <returns>Projection containing CloudEventsResponse.</returns>
    public Angzarr.Projection HandleRequest(Angzarr.EventBook source)
    {
        var response = Project(source);
        return new Angzarr.Projection
        {
            Projector = _name,
            Projection_ = Any.Pack(response, "type.googleapis.com/"),
        };
    }

    /// <summary>
    /// Get the event types this router handles.
    /// </summary>
    /// <returns>List of event type suffixes.</returns>
    public IReadOnlyList<string> EventTypes()
    {
        return _handlers.Select(h => h.Suffix).ToList();
    }

    /// <summary>
    /// Get the subscription configuration for framework registration.
    /// </summary>
    /// <returns>Tuple of (domain, event types).</returns>
    public (string Domain, IReadOnlyList<string> Types) Subscriptions()
    {
        return (_inputDomain, EventTypes());
    }
}

File: client/python/angzarr_client/cloudevents.py
"""CloudEvents support for Angzarr projectors.

CloudEvents projectors transform internal domain events into CloudEvents 1.0 format
for external consumption via HTTP webhooks or Kafka.

OO Pattern (CloudEventsProjector):
    from angzarr_client import CloudEvent, CloudEventsProjector

    @domain("player")
    class PlayerCloudEventsProjector(CloudEventsProjector):
        name = "prj-player-cloudevents"

        def on_player_registered(self, event: PlayerRegistered) -> CloudEvent | None:
            public = PublicPlayerRegistered(display_name=event.display_name)
            data = Any()
            data.Pack(public)
            return CloudEvent(type="com.poker.player.registered", data=data)

Functional Pattern (CloudEventsRouter):
    from angzarr_client import CloudEvent, CloudEventsRouter

    def handle_player_registered(event: PlayerRegistered) -> CloudEvent | None:
        public = PublicPlayerRegistered(display_name=event.display_name)
        data = Any()
        data.Pack(public)
        return CloudEvent(type="com.poker.player.registered", data=data)

    router = (
        CloudEventsRouter("prj-player-cloudevents", "player")
        .on("PlayerRegistered", handle_player_registered)
    )
"""

from __future__ import annotations

import inspect
from abc import ABC
from typing import Callable, Optional

from google.protobuf.any_pb2 import Any
from google.protobuf.message import Message

from .helpers import TYPE_URL_PREFIX
from .proto.angzarr import types_pb2 as types
from .proto.angzarr.cloudevents_pb2 import CloudEvent, CloudEventsResponse
from .router import domain

__all__ = [
    "CloudEvent",
    "CloudEventsProjector",
    "CloudEventsRouter",
    "CloudEventsResponse",
]


class CloudEventsProjector(ABC):
    """Base class for CloudEvents projectors using method naming convention.

    Handler methods are named `on_{event_type}` where event_type is the snake_case
    version of the protobuf message name (e.g., on_player_registered for PlayerRegistered).

    Subclasses must:
    - Set `name` class attribute
    - Use @domain decorator to set input domain
    - Define `on_{event_type}` methods that return CloudEvent or None

    Example:
        @domain("player")
        class PlayerCloudEventsProjector(CloudEventsProjector):
            name = "prj-player-cloudevents"

            def on_player_registered(self, event: PlayerRegistered) -> CloudEvent | None:
                public = PublicPlayerRegistered(display_name=event.display_name)
                data = Any()
                data.Pack(public)
                return CloudEvent(type="com.poker.player.registered", data=data)
    """

    name: str
    _domain: str = None  # Set by @domain decorator

    @property
    def input_domain(self) -> str:
        """Get input domain (from @domain decorator or __init__ argument)."""
        return self._domain

    def __init__(self, name: str = None, input_domain: str = None) -> None:
        """Initialize projector.

        Args:
            name: Projector name (optional, uses class attribute if not provided)
            input_domain: Input domain (optional, uses @domain decorator if not provided)
        """
        if name:
            self.name = name
        if input_domain:
            self._domain = input_domain

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)

        # Skip validation for abstract intermediate classes
        if inspect.isabstract(cls):
            return

    def project(self, source: types.EventBook) -> list[CloudEvent]:
        """Process an EventBook and return CloudEvents.

        Args:
            source: EventBook containing events to process.

        Returns:
            List of CloudEvents for external publishing.
        """
        events = []

        for page in source.pages:
            if not page.HasField("event"):
                continue

            # Extract event type suffix from type_url
            type_url = page.event.type_url
            suffix = type_url.rsplit("/", 1)[-1].rsplit(".", 1)[-1]

            # Convert PascalCase to snake_case for method name
            method_name = f"on_{_pascal_to_snake(suffix)}"

            # Try to call handler method
            handler = getattr(self, method_name, None)
            if handler and callable(handler):
                # Need to unpack the event - try to infer type from type hints
                hints = getattr(handler, "__annotations__", {})
                event_types = [
                    v for k, v in hints.items() if k != "return" and isinstance(v, type)
                ]
                if event_types:
                    event_type = event_types[0]
                    event = event_type()
                    page.event.Unpack(event)
                    result = handler(event)
                    if result is not None:
                        events.append(result)

        return events


class CloudEventsRouter:
    """Functional router for CloudEvents projectors.

    Example:
        def handle_player_registered(event: PlayerRegistered) -> CloudEvent | None:
            public = PublicPlayerRegistered(display_name=event.display_name)
            data = Any()
            data.Pack(public)
            return CloudEvent(type="com.poker.player.registered", data=data)

        router = (
            CloudEventsRouter("prj-player-cloudevents", "player")
            .on("PlayerRegistered", handle_player_registered)
        )
    """

    def __init__(self, name: str, input_domain: str) -> None:
        self.name = name
        self.input_domain = input_domain
        self._handlers: dict[str, tuple[type, Callable]] = {}

    def on(
        self,
        event_suffix: str,
        handler: Callable[[Message], Optional[CloudEvent]],
    ) -> "CloudEventsRouter":
        """Register a handler for an event type.

        Args:
            event_suffix: Event type suffix to match (e.g., "PlayerRegistered")
            handler: Function that takes the event and returns CloudEvent or None

        Returns:
            Self for chaining.
        """
        # Get event type from type hints
        hints = getattr(handler, "__annotations__", {})
        event_types = [
            v for k, v in hints.items() if k != "return" and isinstance(v, type)
        ]
        event_type = event_types[0] if event_types else None
        self._handlers[event_suffix] = (event_type, handler)
        return self

    def project(self, source: types.EventBook) -> list[CloudEvent]:
        """Process an EventBook and return CloudEvents.

        Args:
            source: EventBook containing events to process.

        Returns:
            List of CloudEvents for external publishing.
        """
        events = []

        for page in source.pages:
            if not page.HasField("event"):
                continue

            # Extract event type suffix from type_url
            type_url = page.event.type_url
            suffix = type_url.rsplit("/", 1)[-1].rsplit(".", 1)[-1]

            handler_info = self._handlers.get(suffix)
            if handler_info:
                event_type, handler = handler_info
                if event_type:
                    event = event_type()
                    page.event.Unpack(event)
                    result = handler(event)
                else:
                    # No type hint, pass the Any directly
                    result = handler(page.event)

                if result is not None:
                    events.append(result)

        return events


def _pascal_to_snake(name: str) -> str:
    """Convert PascalCase to snake_case."""
    result = []
    for i, char in enumerate(name):
        if char.isupper() and i > 0:
            result.append("_")
        result.append(char.lower())
    return "".join(result)
